src/librustc_lint/unused.rs

```diff
@@ -464,3 +464,61 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedAllocation {
         }
     }
 }
+
+declare_lint! {
+    pub(super) UNUSED_LABEL,
+    Warn,
+    "warns on unused labels"
+}
+
+#[derive(Clone)]
+pub struct UnusedLabel(pub Vec<ast::Label>);
+
+impl UnusedLabel {
+    pub fn new() -> Self {
+        UnusedLabel(vec![])
+    }
+}
+
+impl LintPass for UnusedLabel {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(UNUSED_LABEL)
+    }
+}
+
+impl EarlyLintPass for UnusedLabel {
+    fn check_expr(&mut self, _: &EarlyContext, expr: &ast::Expr) {
+        match expr.node {
+            ast::ExprKind::While(_, _, Some(label))
+            | ast::ExprKind::WhileLet(_, _, _, Some(label))
+            | ast::ExprKind::ForLoop(_, _, _, Some(label))
+            | ast::ExprKind::Loop(_, Some(label)) => {
+                self.0.push(label);
+            }
+            ast::ExprKind::Break(Some(label), _) | ast::ExprKind::Continue(Some(label)) => {
+                if let Some(index) = self.0.iter().rposition(|&l| l.ident == label.ident) {
+                    self.0.remove(index);
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn check_expr_post(&mut self, ctxt: &EarlyContext, expr: &ast::Expr) {
+        match expr.node {
+            ast::ExprKind::While(_, _, Some(label))
+            | ast::ExprKind::WhileLet(_, _, _, Some(label))
+            | ast::ExprKind::ForLoop(_, _, _, Some(label))
+            | ast::ExprKind::Loop(_, Some(label)) => {
+                if let Some(unused_label) = self.0.pop() {
+                    if label.ident == unused_label.ident {
+                        ctxt.span_lint(UNUSED_LABEL, label.ident.span, "unused label");
+                    } else {
+                        self.0.push(unused_label);
+                    }
```

Comment:
I'm not sure this is correct. If you have multiple nested scopes with labels, `check_expr_post` is going to start going down, right? If this is the case, then the lint will only warn on the bottom-most label. I'm not entirely sure. That being said, I feel this would be easier to follow as a `check_crate_post` and merely iterate through `&self.0` idents and emit the lint.

Reply:
Yeah, this approach definitely has an issue with what you mentioned earlier:
```rust
fn main() {
    'used_many_shadowed: loop {
        'used_many_shadowed: loop {
            if true {
                break 'used_many_shadowed;
            } else {
                break 'used_many_shadowed;
            }
        }
    }
}
```
Should warn on the first `'used_many_shadowed` label, but won't. I've implemented a fix using a `Vec<(Label, bool)>` to keep track of when each `Label` is used instead of just removing it, but honestly it's looking like @petrochenkov 's suggestion to run this check within `librustc_resolve` is probably the way to go.